### PR TITLE
Add `Ast` as top level enum

### PR DESCRIPTION
**Pull request by youknowone**
One of the possible solution of #48

### CODE REVIEW

**Comment by MichaReiser** [comment]
Thanks for working on this and sorry for my late reply. I've been busy with other work and haven't had time to think more about the problem. 

I still think that I would favor the flat structure over the nested one because:

* The nested structure requires two marker bits to distinguish the variants. One at the top level, and another for the sub-union.  That means, the enum is three bytes larger than a flat structure. 
* In my view, the most common access patterns are that we don't care about what the node is OR that we want to test if it is a few very specific nodes. The nested structure requires a nested match except for, in my view, the rare case where you want to match on all statements. 
* I think we can accommodate the "matching on all statements" by introducing the helpers `is_stmt()`, `as_stmt()` and `stmt()` on the flat structure. 


Would you mind if I change the enum to a flat structure? I will also introduce a `AstRef` as part of that work because that's actually the enum that I need first.

File: ast/src/gen/generic.rs [hunk 1]
```diff
@@ -1,6 +1,34 @@
...
+    Alias(Alias<R>),
+    Withitem(Withitem<R>),
+    MatchCase(MatchCase<R>),
+    Pattern(Pattern<R>),
+    TypeIgnore(TypeIgnore<R>),
+}
+impl<R> Node for Ast<R> {
+    const NAME: &'static str = "AST";
+    const FIELD_NAMES: &'static [&'static str] = &[];
+}
```
Comment (MichaReiser) [inline ast/src/gen/generic.rs:30]: What's the reason that the enums implement the `Node` trait. They aren't really nodes and have no fields.

**Comment by youknowone** [comment]
I wish conversion between Ast and other types like Stmt and Expr doesn't require much cost. If we can achieve it with flat structure, that's also fine.

**Comment by MichaReiser** [comment]
> Because current `Ast` enum is the counter part of python side `ast.AST`(each rust enum matches to inheritance from Python), adding a new enum for flat structure seems better if you need one. Before adding the `Ast` type, the top level `ast.AST` was the only missing type from Rust side. On the other hand, adding `is_stmt_if()` to current `Ast` seems also reasonable if it fits your requirements. (probably not if you need to use match)

I'll add it on the Ruff side if you want to keep using the `Ast` node for CPython compatibility

**Pull request merged**